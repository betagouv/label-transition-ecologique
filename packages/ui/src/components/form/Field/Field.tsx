import classNames from 'classnames';

type Props = {
  /** Titre */
  title: string;
  /** Enfant: input, select... */
  children: React.ReactNode;
  /** class donnée au container */
  className?: string;
  /** Complément d'informations */
  hint?: string;
  /** Pour lier à un formulaire */
  htmlFor?: string;
  /** État */
  state?: 'default' | 'info' | 'error' | 'success' | 'warning' | 'disabled';
  /** Message de l'état afficher dessous */
  message?: string;
};

/** Wrapper pour élément de formulaire donnant des informations et un état */
const Field = ({
  title,
  className,
  hint,
  htmlFor,
  state = 'default',
  message,
  children,
}: Props) => {
  return (
    <div className={classNames(`flex-grow flex flex-col gap-3`, className)}>
      <label htmlFor={htmlFor}>
        {/** Title */}
        <div
          className={classNames('font-medium text-grey-8', {
            '!text-grey-5': state === 'disabled',
          })}
        >
          {title}
        </div>
        {/** Hint */}
        {hint !== undefined && (
          <div
            className={classNames('mt-2 text-xs text-grey-6', {
              '!text-grey-5': state === 'disabled',
            })}
          >
            {hint}
          </div>
        )}
      </label>
      {children}
      {message !== undefined && (
        <div
          className={classNames('flex items-center', {
            'text-info-1': state === 'info',
            'text-error-1': state === 'error',
            'text-success-1': state === 'success',
            'text-warning-1': state === 'warning',
          })}
        >
          <span
            className={classNames('flex mr-1 before:h-4 before:w-4', {
              'fr-icon-info-fill': state === 'info',
              'fr-icon-error-fill': state === 'error',
              'fr-icon-success-fill': state === 'success',
              'fr-icon-warning-fill': state === 'warning',
            })}
          />
          <span className="text-xs">{message}</span>
        </div>
      )}
    </div>
  );
};

export default Field;
