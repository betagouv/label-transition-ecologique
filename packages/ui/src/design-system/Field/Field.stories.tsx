import {Meta, StoryObj} from '@storybook/react';

import {Field} from '.';

const meta: Meta<typeof Field> = {
  component: Field,
  decorators: [story => <div className="flex">{story()}</div>],
};

export default meta;

type Story = StoryObj<typeof Field>;

const FakeInput = () => (
  <input
    type="text"
    className="border border-solid border-grey-4 rounded-lg min-h-[3rem] px-4 placeholder:text-sm"
    placeholder="Placeholder"
  />
);

export const Default: Story = {
  args: {title: "Description de l'action"},
  render: args => (
    <div className="flex flex-col gap-6 w-full max-w-sm">
      <Field {...args}>
        <FakeInput />
      </Field>
      <Field {...args} title="Small" small>
        <FakeInput />
      </Field>
    </div>
  ),
};

export const DisabledWithHint: Story = {
  args: {
    title: "Description de l'action",
    state: 'disabled',
    hint: 'Texte description additionnel',
    message: 'Message d’information',
  },
  render: args => (
    <Field {...args}>
      <FakeInput />
    </Field>
  ),
};

export const WithHint: Story = {
  args: {
    title: "Description de l'action",
    hint: 'Texte description additionnel',
    message: 'Message d’information',
  },
  render: args => (
    <Field {...args}>
      <FakeInput />
    </Field>
  ),
};

export const WithInfo: Story = {
  args: {
    title: "Description de l'action",
    state: 'info',
    message: 'Message d’information',
  },
  render: args => (
    <Field {...args}>
      <FakeInput />
    </Field>
  ),
};

export const WithError: Story = {
  args: {
    title: "Description de l'action",
    state: 'error',
    message: "Message d'erreur",
  },
  render: args => (
    <Field {...args}>
      <FakeInput />
    </Field>
  ),
};

export const WithSuccess: Story = {
  args: {
    title: "Description de l'action",
    state: 'success',
    message: 'Message de succès',
  },
  render: args => (
    <Field {...args}>
      <FakeInput />
    </Field>
  ),
};

export const WithWarning: Story = {
  args: {
    title: "Description de l'action",
    state: 'warning',
    message: "Message d'avertissement",
  },
  render: args => (
    <Field {...args}>
      <FakeInput />
    </Field>
  ),
};
