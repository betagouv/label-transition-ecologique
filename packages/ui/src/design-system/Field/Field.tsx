import classNames from 'classnames';
import {Icon} from '@design-system/Icon';

export type FieldState =
  | 'default'
  | 'info'
  | 'error'
  | 'success'
  | 'warning'
  | 'disabled';

export const stateToTextColor: Record<FieldState, string> = {
  info: 'text-info-1',
  error: 'text-error-1',
  success: 'text-success-1',
  warning: 'text-warning-1',
  default: 'text-grey-8',
  disabled: 'text-grey-6',
};

const stateToIcon: Record<FieldState, string> = {
  info: 'information-fill',
  error: 'spam-fill',
  success: 'checkbox-circle-fill',
  warning: 'alert-fill',
  default: '',
  disabled: '',
};

type Props = {
  /** Titre */
  title: string;
  /** Enfant: input, select... */
  children: React.ReactNode;
  /** Pour surcharger les styles du container */
  className?: string;
  /** Complément d'informations */
  hint?: string;
  /** Pour lier le libellé et le champ qu'il contient */
  htmlFor?: string;
  /** Réduit la taille du titre */
  small?: boolean;
} & FieldMessageProps;

/** Wrapper pour élément de formulaire donnant des informations et un état */
export const Field = ({
  title,
  className,
  hint,
  htmlFor,
  state = 'default',
  message,
  children,
  small,
}: Props) => {
  return (
    <div className={classNames(`flex-grow flex flex-col gap-3`, className)}>
      <label htmlFor={htmlFor}>
        {/** Title */}
        <div
          className={classNames('font-medium text-grey-8', {
            'text-grey-5': state === 'disabled',
            'text-sm': small,
          })}
        >
          {title}
        </div>
        {/** Hint */}
        {hint !== undefined && (
          <span
            className={classNames('block mt-2 text-xs text-grey-7', {
              'text-grey-5': state === 'disabled',
            })}
          >
            {hint}
          </span>
        )}
      </label>
      {children}
      <FieldMessage state={state} message={message} />
    </div>
  );
};

export type FieldMessageProps = {
  /** État */
  state?: FieldState;
  /** Message d'état affiché en dessous du champ */
  message?: string;
  /** Pour surcharger les styles du container du message */
  messageClassName?: string;
};

/**
 * Affiche un message additionnel au-dessous d'un champ
 */
export const FieldMessage = ({
  message,
  state = 'default',
  messageClassName,
}: FieldMessageProps) => {
  return (
    message !== undefined && (
      <div
        className={classNames(
          'flex items-center',
          stateToTextColor[state],
          messageClassName
        )}
      >
        {state !== 'disabled' && state !== 'default' && (
          <Icon icon={stateToIcon[state]} size="sm" className="mr-1" />
        )}
        <span className="text-xs">{message}</span>
      </div>
    )
  );
};
