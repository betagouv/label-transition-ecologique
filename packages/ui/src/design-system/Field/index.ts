export * from './Field';
export * from './FieldMessage';
