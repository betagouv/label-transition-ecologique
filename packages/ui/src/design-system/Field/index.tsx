import classNames from 'classnames';
import {Icon} from '@design-system/Icon';

export type FieldState =
  | 'default'
  | 'info'
  | 'error'
  | 'success'
  | 'warning'
  | 'disabled';

export const stateToTextColor: Record<FieldState, string> = {
  info: 'text-info-1',
  error: 'text-error-1',
  success: 'text-success-1',
  warning: 'text-warning-1',
  default: 'text-grey-8',
  disabled: 'text-grey-6',
};

const stateToIcon: Record<FieldState, string> = {
  info: 'information-fill',
  error: 'spam-fill',
  success: 'checkbox-circle-fill',
  warning: 'alert-fill',
  default: '',
  disabled: '',
};

type Props = {
  /** Titre */
  title: string;
  /** Enfant: input, select... */
  children: React.ReactNode;
  /** Pour surcharger les styles du container */
  className?: string;
  /** Complément d'informations */
  hint?: string;
  /** Pour lier le libellé et le champ qu'il contient */
  htmlFor?: string;
  /** État */
  state?: FieldState;
  /** Message d'état affiché en dessous du champ */
  message?: string;
  /** Réduit la taille du titre */
  small?: boolean;
};

/** Wrapper pour élément de formulaire donnant des informations et un état */
export const Field = ({
  title,
  className,
  hint,
  htmlFor,
  state = 'default',
  message,
  children,
  small,
}: Props) => {
  return (
    <div className={classNames(`flex-grow flex flex-col gap-3`, className)}>
      <label htmlFor={htmlFor}>
        {/** Title */}
        <div
          className={classNames('font-medium text-grey-8', {
            'text-grey-5': state === 'disabled',
            'text-sm': small,
          })}
        >
          {title}
        </div>
        {/** Hint */}
        {hint !== undefined && (
          <span
            className={classNames('block mt-2 text-xs text-grey-7', {
              'text-grey-5': state === 'disabled',
            })}
          >
            {hint}
          </span>
        )}
      </label>
      {children}
      {message !== undefined && (
        <div
          className={classNames('flex items-center', stateToTextColor[state])}
        >
          {state !== 'disabled' && state !== 'default' && (
            <Icon icon={stateToIcon[state]} size="sm" className="mr-1" />
          )}
          <span className="text-xs">{message}</span>
        </div>
      )}
    </div>
  );
};
