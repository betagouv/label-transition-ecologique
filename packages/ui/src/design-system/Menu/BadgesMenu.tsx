import {Badge} from '@design-system/Badge';
import classNames from 'classnames';
import {ButtonMenu} from './ButtonMenu';
import {NotificationButton} from '@design-system/Button';

export type BadgeType = {
  /** Label du badge */
  label: string;
  /** Détecte la suppression du badge */
  onClose?: () => void;
};

type BadgesMenuProps = {
  /** Comportement par défaut avec menu flottant sous le bouton, ou comportement custom activé par le click */
  variant?: 'default' | 'custom';
  /** Permet la customisation de l'icône */
  icon?: string;
  /** Liste des badges à afficher */
  badgesList: BadgeType[];
  /** Contenu du menu (uniquement avec le variant `default`) */
  menuContent?: React.ReactNode;
  /** Surcouche sur la classname du menu */
  className?: string;
  /** Détecte le click sur le bouton */
  onClick?: () => void;
  /** Détecte le clear sur l'ensemble des badges */
  onClearBadges?: () => void;
};

export const BadgesMenu = ({
  variant = 'default',
  icon = 'equalizer-fill',
  badgesList,
  menuContent,
  className,
  onClick,
  onClearBadges,
}: BadgesMenuProps) => {
  return (
    <div className={classNames('flex justify-between items-start', className)}>
      <div className="flex gap-2 flex-wrap">
        {/* Liste des badges */}
        {!!badgesList &&
          badgesList.map(badge => (
            <Badge
              key={badge.label}
              title={badge.label}
              size="sm"
              state="standard"
              uppercase={false}
              light
              onClose={badge.onClose}
            />
          ))}

        {/* Badge pour effacer tous les filtres sélectionnés */}
        {!!badgesList && badgesList.length > 0 && !!onClearBadges && (
          <div
            onClick={onClearBadges} // tout le badge est clickable
            className="cursor-pointer"
          >
            <Badge
              title="Supprimer tous les filtres"
              icon="delete-bin-6-line"
              iconPosition="left"
              size="sm"
              state="grey"
              uppercase={false}
              light
              onClose={() => {}}
            />
          </div>
        )}
      </div>

      {/* Bouton par défaut avec ouverture du menu flottant */}
      {variant === 'default' && (
        <ButtonMenu
          icon={icon}
          notificationValue={badgesList?.length || undefined}
          onClick={onClick}
        >
          <div className="flex flex-col gap-4 w-72">{menuContent}</div>
        </ButtonMenu>
      )}

      {/* Bouton classique avec trigger custom au click */}
      {variant === 'custom' && (
        <div className="mx-2 w-fit">
          <NotificationButton
            size="xs"
            icon={icon}
            notificationValue={badgesList?.length || undefined}
            onClick={onClick}
          />
        </div>
      )}
    </div>
  );
};
