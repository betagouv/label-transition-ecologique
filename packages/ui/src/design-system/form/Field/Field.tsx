import classNames from 'classnames';
import {Icon} from '../../icons/Icon';

type State = 'default' | 'info' | 'error' | 'success' | 'warning' | 'disabled';

const stateToColor: Record<State, string> = {
  info: 'text-info-1',
  error: 'text-error-1',
  success: 'text-success-1',
  warning: 'text-warning-1',
  default: '',
  disabled: '',
};

const stateToIcon: Record<State, string> = {
  info: 'information-fill',
  error: 'spam-fill',
  success: 'checkbox-circle-fill',
  warning: 'alert-fill',
  default: '',
  disabled: '',
};

type Props = {
  /** Titre */
  title: string;
  /** Enfant: input, select... */
  children: React.ReactNode;
  /** Pour surcharger les styles du container */
  className?: string;
  /** Complément d'informations */
  hint?: string;
  /** Pour lier le libellé et le champ qu'il contient */
  htmlFor?: string;
  /** État */
  state?: State;
  /** Message d'état affiché en dessous du champ */
  message?: string;
};

/** Wrapper pour élément de formulaire donnant des informations et un état */
const Field = ({
  title,
  className,
  hint,
  htmlFor,
  state = 'default',
  message,
  children,
}: Props) => {
  return (
    <div className={classNames(`flex-grow flex flex-col gap-3`, className)}>
      <label htmlFor={htmlFor}>
        {/** Title */}
        <div
          className={classNames('font-medium text-grey-8', {
            '!text-grey-5': state === 'disabled',
          })}
        >
          {title}
        </div>
        {/** Hint */}
        {hint !== undefined && (
          <span
            className={classNames('block mt-2 text-xs text-grey-6', {
              '!text-grey-5': state === 'disabled',
            })}
          >
            {hint}
          </span>
        )}
      </label>
      {children}
      {message !== undefined && (
        <div className={classNames('flex items-center', stateToColor[state])}>
          <Icon icon={stateToIcon[state]} size="sm" className="mr-1" />
          <span className="text-xs">{message}</span>
        </div>
      )}
    </div>
  );
};

export default Field;
