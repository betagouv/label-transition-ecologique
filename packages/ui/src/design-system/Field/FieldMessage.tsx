import classNames from 'classnames';

import {Icon} from '../Icon';
import {FieldState} from './Field';

export const stateToTextColor: Record<FieldState, string> = {
  info: 'text-info-1',
  error: 'text-error-1',
  success: 'text-success-1',
  warning: 'text-warning-1',
  default: 'text-grey-7',
  disabled: 'text-grey-6',
};

export const stateToIcon: Record<FieldState, string> = {
  info: 'information-fill',
  error: 'spam-fill',
  success: 'checkbox-circle-fill',
  warning: 'alert-fill',
  default: '',
  disabled: '',
};

/**
 * Affiche un message additionnel au-dessous d'un champ de formulaire.
 */
export type FieldMessageProps = {
  /** État */
  state?: FieldState;
  /** Message d'état affiché en dessous du champ */
  message?: string | JSX.Element;
  /** Pour surcharger les styles du container du message */
  messageClassName?: string;
};

export const FieldMessage = ({
  message,
  state = 'default',
  messageClassName,
}: FieldMessageProps) => {
  return (
    message !== undefined && (
      <div
        className={classNames(
          'flex items-center',
          stateToTextColor[state],
          messageClassName
        )}
      >
        {state !== 'disabled' && state !== 'default' && (
          <Icon icon={stateToIcon[state]} size="sm" className="mr-1" />
        )}
        {typeof message === 'string' ? (
          <span className="text-xs">{message}</span>
        ) : (
          message
        )}
      </div>
    )
  );
};
