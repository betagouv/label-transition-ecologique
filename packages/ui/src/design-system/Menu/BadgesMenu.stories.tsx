import {Meta, StoryObj} from '@storybook/react';
import {BadgesMenu} from './BadgesMenu';
import {Field} from '@design-system/Field';
import {
  Option,
  OptionValue,
  SelectMultiple,
  SelectOption,
} from '@design-system/Select';
import {useState} from 'react';

const meta: Meta<typeof BadgesMenu> = {
  component: BadgesMenu,
  argTypes: {
    variant: {
      control: {type: 'select'},
    },
  },
  args: {},
};

export default meta;

type Story = StoryObj<typeof BadgesMenu>;

export const Default: Story = {
  render: args => {
    const [values, setValues] = useState<OptionValue[] | undefined>();

    const options: SelectOption[] = [
      {value: 1, label: 'Test 1'},
      {value: 2, label: 'Test 2'},
      {value: 3, label: 'Test 3'},
    ];

    return (
      <div className="mb-52">
        <BadgesMenu
          {...args}
          badgesList={options
            .filter(opt => values?.includes((opt as Option).value))
            .map(opt => ({
              label: (opt as Option).label,
              onClose: () =>
                setValues(values.filter(v => v !== (opt as Option).value)),
            }))}
          menuContent={
            <div className="flex flex-col gap-4 w-72">
              <Field title="Test">
                <SelectMultiple
                  options={options}
                  values={values}
                  onChange={({values}) => setValues(values)}
                  small
                />
              </Field>
            </div>
          }
          onClearBadges={() => setValues(undefined)}
        />
      </div>
    );
  },
};

export const Custom: Story = {
  render: args => {
    const [badges, setBadges] = useState<string[]>([]);

    return (
      <div className="mb-12">
        <BadgesMenu
          {...args}
          variant="custom"
          badgesList={badges.map(b => ({
            label: b,
          }))}
          onClick={() => {
            const newBadge = `Nouveau badge ${badges.length + 1}`;
            const newBadgesList = [...badges, newBadge];
            setBadges(newBadgesList);
          }}
          onClearBadges={() => setBadges([])}
        />
      </div>
    );
  },
};
