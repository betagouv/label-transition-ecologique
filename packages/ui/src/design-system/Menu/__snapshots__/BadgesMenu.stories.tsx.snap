// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`design-system/Menu/BadgesMenu Custom smoke-test 1`] = `
<div class="mb-12">
  <div class="flex justify-between items-start">
    <div class="flex gap-2 flex-wrap">
    </div>
    <div class="mx-2 w-fit">
      <div class="relative w-fit h-fit">
        <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-white hover:!text-white bg-primary-7 hover:!bg-primary-8 border-primary-7 hover:!border-primary-8 p-2">
          <span class="ri-equalizer-fill font-normal text-center h-4 w-4 text-[1rem] leading-[1rem] fill-white group-hover:fill-white">
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`design-system/Menu/BadgesMenu Default smoke-test 1`] = `
<div class="mb-52">
  <div class="flex justify-between items-start">
    <div class="flex gap-2 flex-wrap">
    </div>
    <div class="mx-2 relative w-fit">
      <div class="relative w-fit h-fit">
        <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-white hover:!text-white bg-primary-7 hover:!bg-primary-8 border-primary-7 hover:!border-primary-8 p-2">
          <span class="ri-equalizer-fill font-normal text-center h-4 w-4 text-[1rem] leading-[1rem] fill-white group-hover:fill-white">
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
